Use strongly typed MediaTypeHeaderValue for content type in action resul...

kichalla opened the pull request:
...ts.

@rynowak @pranavkm

kichalla commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileContentResult.cs:
```diff
@@ -19,13 +20,14 @@ public class FileContentResult : FileResult
         /// </summary>
         /// <param name="fileContents">The bytes that represent the file contents.</param>
-        public FileContentResult([NotNull] byte[] fileContents)
-            : base(contentType: null)
```
If we compare to MVC5, content type was always required a content-type:
http://aspnetwebstack.codeplex.com/SourceControl/latest#src/System.Web.Mvc/FileResult.cs

harshgMSFT replied:
This seems to be like a compatiblity bug fix - should have been dealt with as a separate PR. (Its fine to leave this here for this one, but something to keep in mind in general ).

kichalla commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,22 +28,52 @@ public static class ViewExecutor
-            var wrappedStream = new StreamWrapper(actionContext.HttpContext.Response.Body);
-            var encoding = Encodings.UTF8EncodingWithoutBOM;
-            using (var writer = new StreamWriter(wrappedStream, encoding, BufferSize, leaveOpen: true))
+            using (var writer = new StreamWriter(
```
will revert this formatting

kichalla commented on src/Microsoft.AspNet.Mvc.Core/Controller.cs:
```diff
@@ -357,10 +358,22 @@ public virtual ContentResult Content(string content, string contentType)
+        /// <param name="contentType">The content type (MIME type).</param>
+        /// <returns>The created <see cref="ContentResult"/> object for the response.</returns>
+        [NonAction]
+        public virtual ContentResult Content(string content, MediaTypeHeaderValue contentType)
```
There are lot of helper methods which accept content type as string, should i have corresponding helpers for MediaTypeHeaderValue too? wanted to check first as there are lot of methods to add.

pranavkm replied:
Might be a question for @Eilon \ @danroth27

Eilon replied:
Doesn't seem urgent... the `MediaTypeHeaderValue` ctor seems pretty easy to use. Can always add it later as a convenience if anyone asks for it.

kichalla commented:
ping

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +29,34 @@ public class ContentResult : ActionResult
         {
             var response = context.HttpContext.Response;
 
+            // Encoding property on MediaTypeHeaderValue does not return the exact encoding instance that
```
We shouldn't need to this. If using the encoding set in `ContentType` is incorrect, we have an issue with the way `MediaTypeHeaderValue.Encoding` behaves. cc @Tratcher

Tratcher replied:
I'm planning to remove  MediaTypeHeaderValue.Encoding, it's causing confusion.

pranavkm replied:
:panda_face:

dougbu replied:
@pranavkm is that a happy or sad panda?

pranavkm replied:
Aren't :panda_face: always sad?

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileStreamResult.cs:
```diff
@@ -39,7 +41,7 @@ public FileStreamResult([NotNull] Stream fileStream)
         /// <param name="fileStream">The stream with the file.</param>
         /// <param name="contentType">The Content-Type header of the response.</param>
-        public FileStreamResult([NotNull] Stream fileStream, string contentType)
+        public FileStreamResult([NotNull] Stream fileStream, MediaTypeHeaderValue contentType)
```
In this case, are we responsible for re-encoding `fileStream` if it doesn't match the `Encoding` on `contentType`?

kichalla replied:
I think we are not...this question is applicable to FilePathResult too which uses the following SendFileFeature in some situations.../cc: @Tratcher 

https://github.com/aspnet/Mvc/blob/a9feed4de1eb45bad0750bc868900b20cf842cc7/test/WebSites/FilesWebSite/SendFileMiddleware.cs#L50

Tratcher replied:
From our perspective files just contain bytes. It's up the application to decide what the best encoding is to use with that file and inform the user appropriately.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,22 +28,52 @@ public static class ViewExecutor
+            // is set, so any settings(for example: BOM) on it will be lost when retrieving the value.
+            // In the scenario where the user does not supply encoding, we want to use UTF8 without BOM and
+            // for this reason do not rely on Encoding property.
+            MediaTypeHeaderValue contentTypeHeader;
```
Could we verify what the behavior was in Mvc 5 - what happened when the `ContentEncoding` was set to `Encoding.Utf8` - did we write the BOM?

kichalla replied:
sure..will do...

kichalla replied:
I do not see any effect on the output when ContentEncoding is set to Encoding.Utf8

pranavkm replied:
In which case the right fix for this would be to write a custom `StreamWriter` that doesn't write the BOM and not do whatever we ave here.

kichalla commented:
Updated. /cc: @pranavkm

davidfowl commented:
Are we sure we want to do this?

pranavkm commented:
@davidfowl what's the concern?

kichalla commented:
Updated /cc: @pranavkm @harshgMSFT @rynowak 

In the newest iteration, I added a `ResponseEncodingWrapper` which ignores the preamble of the original encoding(except the UTF16 little endian which we explicitly set to honor https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Common/Encodings.cs#L20)

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +31,47 @@ public class ContentResult : ActionResult
             else
             {
-                contentTypeHeader = new MediaTypeHeaderValue(ContentType);
+                if(string.IsNullOrEmpty(contentTypeHeader.Charset))
```
Can this not go away?

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +31,47 @@ public class ContentResult : ActionResult
             {
-                contentTypeHeader = new MediaTypeHeaderValue("text/plain");
+                contentTypeHeader = DefaultContentType;
+                encoding = Encodings.UTF8EncodingWithoutBOM;
```
`Encoding.Utf8` should suffice.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -11,11 +13,14 @@ namespace Microsoft.AspNet.Mvc
+        public string Content { get; set; }
 
-        public string ContentType { get; set; }
+        public MediaTypeHeaderValue ContentType { get; set; }
```
While you're touching these files, could you add docs?

pranavkm replied:
Bump.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -46,7 +60,7 @@ public class ContentResult : ActionResult
             if (Content != null)
             {
-                await response.WriteAsync(Content, contentTypeHeader.Encoding);
+                await response.WriteAsync(Content, new ResponseEncodingWrapper(encoding));
```
I think it should be fine to flow the encoding through in this case

kichalla replied:
:+1: ...when used with StreamWriter the preamble is being generated...will change it...

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +31,26 @@ public class ContentResult : ActionResult
+                {
+                    // 1. Do not modify the user supplied content type
+                    // 2. Parse here to handle parameters apart from charset
+                    contentTypeHeader = MediaTypeHeaderValue.Parse(contentTypeHeader.ToString());
```
I don't fully follow - why are we `ToString`ing this?

kichalla replied:
we do not want to modify the instance of MediaTypeHeaderValue value that the user supplied to ContentType property and also there could be additional parameters...test: https://github.com/aspnet/Mvc/pull/2450/files#diff-3be60edcf4b71964bc2ec1b25a0984fbR99

Tratcher replied:
It's a .Clone hack. See https://github.com/aspnet/HttpAbstractions/issues/176

pranavkm replied:
:+1:

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FilePathResult.cs:
```diff
@@ -27,15 +28,15 @@ public class FilePathResult : FileResult
         /// <param name="contentType">The Content-Type header of the response.</param>
-        public FilePathResult([NotNull] string fileName)
-            : base(contentType: null)
+        public FilePathResult([NotNull] string fileName, string contentType)
```
Doesn't this remove the overload with `fileName`

harshgMSFT replied:
again use `[NotNull]` if this is required.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FilePathResult.cs:
```diff
@@ -27,15 +28,15 @@ public class FilePathResult : FileResult
-        public FilePathResult([NotNull] string fileName)
-            : base(contentType: null)
+        public FilePathResult([NotNull] string fileName, string contentType)
+            : this(fileName, new MediaTypeHeaderValue(contentType))
```
Will this throw if `contentType` is null?

kichalla replied:
yes, it does

pranavkm replied:
Maybe throw a `[NotNull]` on the argument instead?

pranavkm replied:
bump.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileStreamResult.cs:
```diff
@@ -23,13 +24,14 @@ public class FileStreamResult : FileResult
-            : base(contentType: null)
+        /// <param name="contentType">The Content-Type header of the response.</param>
+        public FileStreamResult([NotNull] Stream fileStream, string contentType)
+            : this(fileStream, new MediaTypeHeaderValue(contentType))
```
Again, is this intentional - the single overload ctor disappeared.

kichalla replied:
yes it is intentional https://github.com/aspnet/Mvc/pull/2450/files#r29091219

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ResponseEncodingWrapper.cs:
```diff
@@ -0,0 +1,55 @@
+    /// <summary>
+    /// Encoding wrapper which makes preamble to be not required.
+    /// </summary>
+    internal class ResponseEncodingWrapper : Encoding
```
If we remove the usage from `ContentResult`, then make this a private type in `ViewExecutor`.

pranavkm commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs:
```diff
@@ -73,45 +79,60 @@ public class ContentResultTest
-                Content = "Test Content",
-                ContentType = "some-type",
-                ContentEncoding = null
+                Content = "abcd",
```
Maybe have a string with some Utf8 chars so we can differentiate between ASCII and UTF8 in the test

pranavkm commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/FileContentResultTest.cs:
```diff
@@ -45,5 +46,29 @@ public void Constructor_SetsFileContents()
         }
+
+        [Fact]
+        public async Task SetsSuppliedContentTypeAndEncoding()
```
`ExecuteResultAsync_SetsSuppliedContentTypeAndEncoding`

pranavkm commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/FilePathResultTest.cs:
```diff
@@ -112,6 +113,37 @@ public void Constructor_SetsFileName()
+
+            // Assert
+            Assert.NotNull(httpContext.Response.Body);
+            var contents = await new StreamReader(httpContext.Response.Body).ReadToEndAsync();
```
Maybe do `var contents = Encoding.ASCII.GetString(bodyStream.ToArray());`

kichalla replied:
:+1:

pranavkm commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/FileResultTest.cs:
```diff
@@ -22,7 +22,7 @@ public void Constructor_SetsContentType()
 
             // Assert
-            Assert.Equal("text/plain", result.ContentType);
+            Assert.Equal("text/plain", result.ContentType?.ToString());
```
`result.ContentType` cannot be null. Drop the `?.`

pranavkm commented:
:watch:

kichalla commented:
Updated /cc: @pranavkm

ajaybhargavb commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/FilePathResultTest.cs:
```diff
@@ -12,6 +12,7 @@
 using Moq;
 using Xunit;
 using System.Text;
+using Microsoft.Net.Http.Headers;
```
Sort.

pranavkm replied:
:100:

ajaybhargavb commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs:
```diff
@@ -1,11 +1,14 @@
-﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+﻿
```
Blank line.

pranavkm commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/FilePathResultTest.cs:
```diff
@@ -112,6 +113,36 @@ public void Constructor_SetsFileName()
+        {
+            // Arrange
+            var expectedContentType = "text/foo; charset=us-ascii";
+            // path will be C:/.../TestFiles/FilePathResultTestFile.txt
```
Comment's out of date.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,22 +29,44 @@ public static class ViewExecutor
+
+            var wrappedStream = new StreamWrapper(response.Body);
+
+            using (var writer = new StreamWriter(
```
Could you add a comment here calling out why we have a wrapper for the encoding?

pranavkm commented:
:watch: mostly done. There's a couple of un-addressed items. Would also like @rynowak to have a look at this?

kichalla commented:
Updated. /cc: @pranavkm @rynowak @ajaybhargavb

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -11,11 +13,17 @@ namespace Microsoft.AspNet.Mvc
+        };
 
-        public Encoding ContentEncoding { get; set; }
+        public string Content { get; set; }
```
Docs for this.

pranavkm commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,22 +29,47 @@ public static class ViewExecutor
+
+            var wrappedStream = new StreamWrapper(response.Body);
+
+            // StreamWriter writes the preamble or BOM bytes to the response if the encoding requires it.
```
:100:

pranavkm commented:
:shipit:. Try and get @rynowak to look at this before you merge it in.

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -11,11 +13,17 @@ namespace Microsoft.AspNet.Mvc
+            Encoding = Encoding.UTF8
+        };
 
-        public Encoding ContentEncoding { get; set; }
```
Are we removing this because ContentEncoding is included in `MediaTypeHeaderValue`?

kichalla replied:
yeah, right...(as a side note...this property name is just confusing...as per Http this would represent the compression encoding like gzip, deflate)..

rynowak replied:
:+1:

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +34,26 @@ public class ContentResult : ActionResult
             {
-                contentTypeHeader = new MediaTypeHeaderValue("text/plain");
+                contentTypeHeader = DefaultContentType;
+                encoding = DefaultContentType.Encoding;
```
Is this same as what existing code does? what is the default encoding for MediaTypeHeaderValue("text/plain") ? Is it UTF8?

kichalla replied:
yeah, right...its UTF8, but we are not specifying the BOM setting as it has no effect in this case, for example when writing content using StreamWriter it writes out the BOM if the encoding requires it, but we are using the following `response.WriteAsync` api which does not use StreamWriter...

https://github.com/aspnet/HttpAbstractions/blob/dev/src/Microsoft.AspNet.Http.Abstractions/HttpResponseWritingExtensions.cs#L39

rynowak replied:
We should use the correct encoding (without BOM) even if it doesn't have an impact. 

This breaks the cycle of people have to ask why we're using plain UTF8 and you having to explain it again.

harshgMSFT replied:
:+1: with what @rynowak  said.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -11,11 +13,17 @@ namespace Microsoft.AspNet.Mvc
-        public string Content { get; set; }
+        private readonly MediaTypeHeaderValue DefaultContentType = new MediaTypeHeaderValue("text/plain")
+        {
+            Encoding = Encoding.UTF8
```
We've defined a static class with these (`Encodings`) - you want to use UTF8 **without BOM**

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +34,26 @@ public class ContentResult : ActionResult
+                    // 1. Do not modify the user supplied content type
+                    // 2. Parse here to handle parameters apart from charset
+                    contentTypeHeader = MediaTypeHeaderValue.Parse(contentTypeHeader.ToString());
+                    contentTypeHeader.Encoding = Encoding.UTF8;
```
If you're creating a `DefaultMediaType` then also create a `DefaultEncoding`

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ContentResult.cs:
```diff
@@ -26,17 +34,26 @@ public class ContentResult : ActionResult
+                encoding = contentTypeHeader.Encoding;
             }
 
-            contentTypeHeader.Encoding = ContentEncoding ?? Encodings.UTF8EncodingWithoutBOM;
```
Please don't change the code to be less correct :laughing:

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileContentResult.cs:
```diff
@@ -19,13 +20,14 @@ public class FileContentResult : FileResult
-        public FileContentResult([NotNull] byte[] fileContents)
-            : base(contentType: null)
+        /// <param name="contentType">The Content-Type header of the response.</param>
+        public FileContentResult([NotNull] byte[] fileContents, string contentType)
```
`[NotNull]`

rynowak replied:
Please review all of these and add `[NotNull]`. The idea is to throw as close as possible to where the mistake occurred.

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileContentResult.cs:
```diff
@@ -35,7 +37,7 @@ public FileContentResult([NotNull] byte[] fileContents)
         /// <param name="fileContents">The bytes that represent the file contents.</param>
         /// <param name="contentType">The Content-Type header of the response.</param>
-        public FileContentResult([NotNull] byte[] fileContents, string contentType)
+        public FileContentResult([NotNull] byte[] fileContents, MediaTypeHeaderValue contentType)
```
`[NotNull]`

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -14,7 +17,10 @@ namespace Microsoft.AspNet.Mvc
+        private static readonly MediaTypeHeaderValue DefaultContentType = new MediaTypeHeaderValue("text/html")
+        {
+            Encoding = Encoding.UTF8
+        };
```
same comments here

rynowak commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,22 +29,47 @@ public static class ViewExecutor
+                wrappedStream,
+                new ResponseEncodingWrapper(encoding),
+                BufferSize,
+                leaveOpen: true))
```
I don't understand why we're doing any of this instead of just using the correct encoding.

rynowak commented:
:watch:

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FilePathResult.cs:
```diff
@@ -46,7 +47,7 @@ public FilePathResult([NotNull] string fileName)
         /// path. Relative and virtual paths are not supported.</param>
         /// <param name="contentType">The Content-Type header of the response.</param>
-        public FilePathResult([NotNull] string fileName, string contentType)
+        public FilePathResult([NotNull] string fileName, MediaTypeHeaderValue contentType)
```
`[NotNull]`

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/FileResult.cs:
```diff
@@ -24,15 +24,25 @@ public abstract class FileResult : ActionResult
+        /// Gets the <see cref="MediaTypeHeaderValue"/> representing the Content-Type header of the response.
         /// </summary>
-        public string ContentType { get; set; }
+        public MediaTypeHeaderValue ContentType { get; }
```
Is it intentional to make this readonly property ?

Eilon replied:
Yeah I think it's intentional because the underlying property on the `MediaTypeHeaderValue` is get-only. It's passed into the ctor of that type and is read-only. So this is a minor breaking change from MVC 5 (or Web API 2 or something), but we discussed on Friday and it's fine.

harshgMSFT replied:
:ok:

harshgMSFT commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -14,7 +17,10 @@ namespace Microsoft.AspNet.Mvc
-        private const string ContentType = "text/html; charset=utf-8";
+        private static readonly MediaTypeHeaderValue DefaultContentType = new MediaTypeHeaderValue("text/html")
+        {
+            Encoding = Encoding.UTF8
```
Same comments as earlier : Use UTF8 Without BOM and use the static helper we have.

harshgMSFT commented on test/Microsoft.AspNet.Mvc.Core.Test/ActionResults/ContentResultTest.cs:
```diff
@@ -73,45 +77,65 @@ public class ContentResultTest
-        public async Task ContentResult_Response_BadContentType_ThrowsFormatException()
+        [Theory]
+        [MemberData(nameof(ContentResultContentTypeData))]
+        public async Task ContentResult_SetContentTypeAndEncoding(
```
Doesnt tell me what are we testing here?

kichalla replied:
I think the name is clear enough?

harshgMSFT replied:
`ContentResult_ExecuteResultAsync_SetContentTypeAndEncoding_OnResponse`

harshgMSFT commented:
:watch:

kichalla commented on src/Microsoft.AspNet.Mvc.Core/ActionResults/ViewExecutor.cs:
```diff
@@ -23,21 +29,43 @@ public static class ViewExecutor
-            actionContext.HttpContext.Response.ContentType = contentType;
-            var wrappedStream = new StreamWrapper(actionContext.HttpContext.Response.Body);
-            var encoding = Encodings.UTF8EncodingWithoutBOM;
             using (var writer = new StreamWriter(wrappedStream, encoding, BufferSize, leaveOpen: true))
```
As discussed, filed: https://github.com/aspnet/Mvc/issues/2496

kichalla commented:
Updated as per our discussion. /cc: @pranavkm @rynowak @harshgMSFT

rynowak commented:
:shipit:

harshgMSFT commented:
Sorry for commenting on the commit rather than on PR. :shipit: after addressing the comments.
